### ISSUE TITLE
Depth of Field: incorrect masking with transparent materials

### ISSUE THREAD

**Issue by davcri** [pmndrs/postprocessing#601]
### Description of the bug

Rendered image isn't correct when using DoF with transparent materials.

### To Reproduce

https://2zcyyg.csb.app/

### Screenshots

https://github.com/pmndrs/postprocessing/assets/6860637/98358d2b-d3bc-4e44-9061-0188685eb267

**Comment by davcri** [comment]
The example is a bit misleading since I'm using a `MeshPhysicalMaterial` with transmission (`transparency=true` has no effect).

However I get the same result by using a `MeshStandardMaterial` with transparency and opacity < 1.0 (https://codesandbox.io/p/sandbox/react-postprocessing-dof-isssue-601-meshstandardmaterial-p4grw6?file=%2Fsrc%2FApp.js%3A65%2C74)

---

The main issue is that the object writes to the depth buffer and this avoids blurring the background, even if it should be.
![image](https://github.com/pmndrs/postprocessing/assets/6860637/65cc514b-c9a5-424f-80be-ced17499f128)

One workaround would be increasing `roughness` on the MeshPhysicalMaterial, however I think this isn't still a realistic effect (since reflections would be affected).

**Comment by davcri** [comment]
It would work for envmap backgrounds but it wouldn't work with proper meshes (which unfortunately is what i have in the project I was working on while I found out the bug).

I took this image from a Unity thread, hopefully this clarifies what I'd like to achieve:

![image](https://github.com/pmndrs/postprocessing/assets/6860637/923c802a-1b2d-493e-8bbb-10d8ca6ec7ef)

PS: thanks again for all the replies @vanruesc , sorry for being pedantic but I didn't find anything online regarding ThreeJS and I'd like to find a fix for this (probably not so common) problem.

**Comment by vanruesc** [comment]
> The main issue is that the object writes to the depth buffer and this avoids blurring the background, even if it should be.

The `DepthOfFieldEffect` blurs objects based on depth and that's working as intended in this case. We can't see the actual background through the transmissive object - that's effectively just its texture.

The deeper problem with transmission is that the renderer [renders the scene to a transmission texture](https://github.com/mrdoob/three.js/blob/cfdd878cf74857aef26459eb82c68dd75f097c72/src/renderers/WebGLRenderer.js#L1430) in an internal process that can't be modified. If this was customizable, we might be able to apply DoF to the transmission texture that is later read by the transmissive object. We'd then have to render DoF again for the fully rendered scene, but we'd end up blurring the transmissive object twice...

I don't know if there's a solution to this.

**Comment by davcri** [comment]
Yeah you're right about MeshPhysicalMaterial, however the problem affects also other materials with transparency, correct? 

 > I don't know if there's a solution to this.
 
Me neither, but I'd lke to investigate about the required changes in ThreeJS and Postprocessing in order to have a proper support for DoF and transparent objects out of the box.
Do you know anyone in the community that could give some hint?

---

In the meantime I updated the first comment with:
1. a simplified explanation of the problem/workarounds
2. a [codesanbox with a simulation](https://codesandbox.io/p/devbox/threejs-pp-forked-nqvlph?file=%2Fsrc%2Findex.js%3A154%2C1) of the desired effect (naive approach that doesn't take depth into account)

In the simulation I did something like this:
```js
composer.addPass(new RenderPass(sceneWithOpaqueObjects, camera));
composer.addPass(new EffectPass(camera, new DepthOfFieldEffect(camera, {...})));
composer.addPass(new CopyPass(opaqueRenderTarget));
composer.addPass(new ClearPass(true, true));
composer.addPass(new RenderPass(sceneWithTransparentCube, camera));
composer.addPass(new CopyPass(transparentRenderTarget));
// render transparent texture on top of the opaque texture
composer.addPass(new EffectPass(camera, new CustomEffect(transparentRenderTarget.texture, opaqueRenderTarget.texture)));
```

**Comment by vanruesc** [comment]
> however the problem affects also other materials with transparency, correct?

Yes, it applies to all transparent meshes. However, transmission is a different problem compared to true transparency. And transparency comes with different blend modes: alpha-blend, alpha-test and alpha-hash. 
- Alpha-Blend has the same issue as you've encountered with transmission.
- Alpha-test is not affected because it's just a cutout based on alpha.
- Alpha-hash with TAA might just work, but I haven't tried that since we don't have a good `TAAEffect` yet.

> Do you know anyone in the community that could give some hint?

No, and I don't think anyone would have a definitive answer. Probably relevant: https://forum.unity.com/threads/transparent-rendering-with-dof.784823/

**Comment by vanruesc** [comment]
I feel like this is one of those issues that are bound to stay open forever since there's no clear solution. Would it be ok if I convert it into a discussion instead?

**Comment by davcri** [comment]
@vanruesc ok for me
